pkg/controller/flow/reconcile.go

```diff
@@ -0,0 +1,360 @@
+/*
+Copyright 2018 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package flow
+
+import (
+	"context"
+	"fmt"
+	"log"
+
+	"github.com/golang/glog"
+	channelsv1alpha1 "github.com/knative/eventing/pkg/apis/channels/v1alpha1"
+	feedsv1alpha1 "github.com/knative/eventing/pkg/apis/feeds/v1alpha1"
+	v1alpha1 "github.com/knative/eventing/pkg/apis/flows/v1alpha1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+// TODO: This should come from a configmap
+const defaultBusName = "stub"
+
+// What field do we assume Object Reference exports as a resolvable target
+const targetFieldName = "domainInternal"
+
+const (
+	// SuccessSynced is used as part of the Event 'reason' when a Flow is synced
+	SuccessSynced = "Synced"
+
+	// MessageResourceSynced is the message used for an Event fired when a Flow
+	// is synced successfully
+	MessageResourceSynced = "Flow synced successfully"
+)
+
+var (
+	flowControllerKind = v1alpha1.SchemeGroupVersion.WithKind("Flow")
+)
+
+// Reconcile compares the actual state with the desired, and attempts to
+// converge the two. It then updates the Status block of the Flow resource
+// with the current status of the resource.
+func (r *reconciler) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	glog.Infof("Reconciling flow %v", request)
+	flow := &v1alpha1.Flow{}
+	err := r.client.Get(context.TODO(), request.NamespacedName, flow)
+
+	if errors.IsNotFound(err) {
+		glog.Errorf("could not find flow %v\n", request)
+		return reconcile.Result{}, nil
+	}
+
+	if err != nil {
+		glog.Errorf("could not fetch Flow %v for %+v\n", err, request)
+		return reconcile.Result{}, err
+	}
+
+	original := flow.DeepCopy()
+
+	flow.Status.InitializeConditions()
+
+	// Reconcile this copy of the Flow and then write back any status
+	// updates regardless of whether the reconcile error out.
+	err = r.reconcile(flow)
+	if equality.Semantic.DeepEqual(original.Status, flow.Status) {
+		// If we didn't change anything then don't call updateStatus.
+		// This is important because the copy we loaded from the informer's
+		// cache may be stale and we don't want to overwrite a prior update
+		// to status with this stale state.
```

Comment:
This update handling is different from the feed controller and I'm curious why. Did you run into issues, or is there a bug in the feed controller's update handling?

The feed controller does this:
1. Get the feed from the cache (feed A)
2. Reconcile feed A
3. Get the feed from the cache (feed B)
4. If the reconciled feed A status is equal to the just-retrieved feed B status, stop and do not update.
5. Copy feed A status into feed B.
6. Update feed B. This update will fail if the feed has been updated by something else since feed B was retrieved (because the apiserver will reject the feed B resourceVersion).

The flow controller does this:
1. Get the flow from the cache (flow A)
2. Copy flow A to original
3. Reconcile flow A
4. If the reconciled flow A status is equal to original status, stop and do not update.
5. Get the flow from the cache (flow B)
6. Copy flow A status into flow B
7. Update flow B. This update will fail if the feed has been updated by something else since feed B was retrieved (because the apiserver will reject the feed B resourceVersion).